README.md
# kcopula

## Overview
kcopula provides the bivariate K-copula by Wollschläger and Schäfer (2016).
It provides two functions:
* `pkcopula()` gives the distribution function of the bivariate K-copula.
* `dkcopula()` gives the density of the bivariate K-copula.

## Installation
Install release version from CRAN:
```r
install.packages("kcopula")
```
Install development version from GitHub:
```r
# install.packages("devtools")
devtools::install_github("mlkremer/kcopula")
```

## Usage
This example plots the bivariate K-copula density and distribution function.
```r
library(kcopula)

## Parameters
b <- .05   # step size
u <- seq(b, 1 - b, b)
v <- u
rho <- .2
N <- 4

## K-copula CDF
pkcopula(.5, .5, rho, N)

## Plot full K-copula CDF
kcopula_cdf <- pkcopula(u, v, rho, N, output = "matrix")
persp(u, v, kcopula_cdf, zlim = c(0, 1), xlab = "\n\n u", ylab = "\n\n v",
      zlab = "\n\n CDF", theta = 30, phi = 30, ticktype = "detailed")

## K-copula PDF
dkcopula(.5, .5, rho, N)

## Plot full K-copula PDF
kcopula_pdf <- dkcopula(u, v, rho, N, output = "matrix")
persp(u, v, kcopula_pdf, zlim = c(0, max(kcopula_pdf)), xlab = "\n\n u", ylab = "\n\n v",
      zlab = "\n\n PDF", theta = 30, phi = 30, ticktype = "detailed")
```

## References
Wollschläger, M. and Schäfer, R. (2016). Impact of nonstationarity on estimating 
and modeling empirical copulas of daily stock returns. 
*Journal of Risk*, 19(1):1--23. https://doi.org/10.21314/JOR.2016.342. 
SSRN version: https://ssrn.com/abstract=3533903.

Chetalova, D., Wollschläger, M., and Schäfer, R. (2015). 
Dependence structure of market states.
*Journal of Statistical Mechanics: Theory and Experiment*, 2015(8):P08012.
https://doi.org/10.1088/1742-5468/2015/08/P08012.
SSRN version: https://ssrn.com/abstract=3533951.


NEWS.md
# kcopula 0.1.0 (February 4, 2020)
* First release.


test_script.R
library(kcopula)
invisible(lapply(c("plot3D", "rgl", "plotly"), library, character.only = T))

## Parameters
b <- .05   # step size
u <- seq(b, 1 - b, b)   # seq(b/2, 1 - b/2, b)
v <- u
rho <- .2
N <- 4

## K-copula PDF
dkcopula(.5, .5, rho, N)
kcopula_pdf <- dkcopula(u, v, rho, N, "matrix")   # takes 7 seconds (system.time())

# 3D plots using different packages
# graphics
persp(u, v, kcopula_pdf, zlim = c(0, max(kcopula_pdf)), xlab = "\n\n u", ylab = "\n\n v",
      zlab = "\n\n PDF", theta = 30, phi = 30, ticktype = "detailed")
# plot3D
persp3D(u, v, kcopula_pdf, zlim = c(0, max(kcopula_pdf)), xlab = "\n\n u", ylab = "\n\n v",
        zlab = "\n\n PDF", ticktype = "detailed")
# rgl
persp3d(u, v, kcopula_pdf)
# plotly
plot_ly(x = u, y = u, z = kcopula_pdf) %>% add_surface()


## K-copula CDF
pkcopula(.5, .5, rho, N)  # takes 7 or 94 seconds
kcopula_cdf <- pkcopula(u, v, rho, N, "matrix")   # takes 7 seconds

persp(u, v, kcopula_cdf, zlim = c(0, 1), xlab = "\n\n u", ylab = "\n\n v",
      zlab = "\n\n CDF", theta = 30, phi = 30, ticktype = "detailed")


R/kcop.R
#' The Bivariate K-Copula
#'
#' Density and distribution function of the bivariate K-copula
#' by Wollschläger and Schäfer (2016).
#'
#' @param u,v Numeric vectors with values in \eqn{[0, 1]}.
#' @param c Numeric; Pearson correlation coefficient in \eqn{[-1, 1]}.
#' @param N Numeric; inverse fluctuation strength of correlations around
#'     their average \code{c}, \eqn{N>0}. The larger \code{N} the smaller the
#'     fluctuations around \code{c}, and vice versa.
#' @param output Character; output as "vector" (default) for single values of
#'     the K-copula, or "matrix" for the full K-copula.
#' @param method Character; method to be used for
#'     \code{pkcopula(..., output = "vector")}.
#'     If \code{method = "interpolate"} (default), values are computed by
#'     interpolating the bivariate K-copula \emph{distribution function}
#'     (computationally faster); returns \code{NA}, if \code{u, v} are out of
#'     range (here: outside of \eqn{[.025, .975]}).
#'     If \code{method = "integrate"}, values are computed by integrating
#'     the bivariate K-copula \emph{density} (computationally slower).
#'
#' @return \code{dkcopula} gives the density (PDF), \code{pkcopula}
#'     gives the distribution function (CDF) of the bivariate K-copula.
#'
#' @inherit kcopula author references
#'
#' @keywords distribution multivariate
#'
#' @name kcop
#'
#' @examples
#' ## Parameters
#' u <- seq(.05, .95, .05)
#' v <- u
#' rho <- .2
#' N <- 4
#'
#' ## K-copula PDF
#' dkcopula(.5, .5, rho, N)
#'
#' \donttest{
#' ## Plot full K-copula PDF
#' kcopula_pdf <- dkcopula(u, v, rho, N, output = "matrix")
#' persp(u, v, kcopula_pdf)
#'
#' ## K-copula CDF
#' pkcopula(.5, .5, rho, N)
#'
#' ## Plot full K-copula CDF
#' kcopula_cdf <- pkcopula(u, v, rho, N, output = "matrix")
#' persp(u, v, kcopula_cdf)
#' }
#'
NULL

#' @rdname kcop
#' @importFrom pracma interp2
#' @export
#'
pkcopula <- function(u, v, c, N, output = "vector", method = "interpolate") {

  cop_cdf <- Vectorize(function(u, v, c, N) {
    Fj(Fmi(u, N), Fmi(v, N), c, N)
  })

  if (output == "vector") {

    if (method == "interpolate") {
      b <- .05   # step size
      x <- seq(b/2, 1 - b/2, b)
      y <- x
      Z <- (apply(apply(dkcopula(x, y, c, N, "matrix"), 2, cumsum), 1, cumsum)
            * diff(x)[1] * diff(y)[1])
      pracma::interp2(x, y, Z, u, v)

    } else if (method == "integrate") {
      pracma::integral2(dkcopula, 1e-3, u, 1e-3, v, c = c, N = N)$Q
      # cubature::hcubature(dkcopula, c(1e-3, 1e-3), c(u, v), c = c, N = N)$integral
      # cop_cdf(u, v, c, N)   # does not work for all parameter sets

    }

  } else if (output == "matrix") {
    (apply(apply(dkcopula(u, v, c, N, "matrix"), 2, cumsum), 1, cumsum)
     * diff(u)[1] * diff(v)[1])
  }
}

#' @rdname kcop
#' @export
#'
dkcopula <- function(u, v, c, N, output = "vector") {

  cop_pdf <- Vectorize(function(u, v, c, N) {
    fj(Fmi(u, N), Fmi(v, N), c, N) / (fm(Fmi(u, N), N) * fm(Fmi(v, N), N))
  })

  if (output == "vector") {
    cop_pdf(u, v, c, N)

  } else if (output == "matrix") {
    outer(u, v, cop_pdf, c, N)
  }
}


R/helper_functions.R
## Marginal CDF of the K-distribution
Fm <- Vectorize(function(x, N) {
  integrand <- function(z, x, N) {
    z^(N/2-1)*exp(-z)*pnorm(x*sqrt(N/(2*z)))
    }
  1/(gamma(N/2))*integrate(integrand, 0, Inf, x = x, N = N)$value
})

## Marginal PDF of the K-distribution
fm <- Vectorize(function(x, N) {
  integrand <- function(z, x, N) {
    z^(N/2-1)*exp(-z)*exp(-N/(4*z)*x^2)/sqrt(4*pi*z/N)
    }
  1/(gamma(N/2))*integrate(integrand, 0, Inf, x = x, N = N)$value
})

## Inverse marginal CDF of the K-distribution
Fmi <- Vectorize(function(u, N) {
  x <- seq(-4, 4, .1)  # sufficient for m \in [0.1, 500] (i.e. always)
  y <- Fm(x, N)  # sapply(x, function(a) {Fm(a, N)})
  approx(y, x, xout = u)$y
})

## Joint CDF of the K-distribution
#' @importFrom pracma integral3
Fj <- Vectorize(function(x, y, c, N) {
  integrand <- function(z, x, y, c, N) {
    1/(gamma(N/2))*z^(N/2-1)*exp(-z)/sqrt(1-c^2)*N/(4*pi*z)*exp(-N/(4*z)*(x^2-2*c*x*y+y^2)/(1-c^2))
  }
  a <- x
  b <- y
  integrand_tf <- function(z, x, y, c, N) {
    x_t <- a - (1-x)/x
    y_t <- b - (1-y)/y
    z_t <- z/(1-z)
    integrand(z_t, x_t, y_t, c = c, N = N)*(1/(1-z)^2)*(1/x^2)*(1/y^2)
  }
  pracma::integral3(integrand_tf, 0, 1, 0, 1, 0, 1, c = c, N = N)#$Q
  # cubature::hcubature(integrand_tf, c(0, 0, 0), c(1, 1, 1), c = c, N = N)$integral
})

## Joint PDF of the K-distribution
fj <- Vectorize(function(x, y, c, N) {
  integrand <- function(z, x, y, c, N) {
    z^(N/2-1)*exp(-z)/sqrt(1-c^2)*N/(4*pi*z)*exp(-N/(4*z)*(x^2-2*c*x*y+y^2)/(1-c^2))
    }
  1/(gamma(N/2))*integrate(integrand, 0, Inf, x = x, y = y, c = c, N = N)$value
  # 1/(gamma(N/2))*pracma::integral(integrand, 0, Inf, x = x, y = y, c = c, N = N)$Q
  # 1/(gamma(N/2))*cubature::hcubature(integrand, 0, Inf, x = x, y = y, c = c, N = N)$integral
})


R/kcopula.R
#' kcopula: The Bivariate K-Copula
#'
#' kcopula provides the bivariate K-copula by Wollschläger and Schäfer (2016).
#'
#' @details kcopula provides two functions:
#'     \itemize{
#'     \item \code{\link{pkcopula}} gives the distribution function of the
#'     bivariate K-copula.
#'     \item \code{\link{dkcopula}} gives the density of the bivariate K-copula.
#'     }
#'
#' @author Marcel Kremer, \email{marcel.kremer@@uni-due.de}
#'
#' @references
#'     Wollschläger, M. and Schäfer, R. (2016). Impact of nonstationarity on
#'     estimating and modeling empirical copulas of daily stock returns.
#'     \emph{Journal of Risk}, 19(1):1--23.
#'     \url{https://doi.org/10.21314/JOR.2016.342}.
#'     SSRN version: \url{https://ssrn.com/abstract=3533903}.
#' @references
#'     Chetalova, D., Wollschläger, M., and Schäfer, R. (2015).
#'     Dependence structure of market states.
#'     \emph{Journal of Statistical Mechanics: Theory and Experiment},
#'     2015(8):P08012. \url{https://doi.org/10.1088/1742-5468/2015/08/P08012}.
#'     SSRN version: \url{https://ssrn.com/abstract=3533951}.
#'
#' @seealso Useful links:
#'     \itemize{
#'     \item \url{https://github.com/mlkremer/kcopula}
#'     \item Report bugs at \url{https://github.com/mlkremer/kcopula/issues}
#'     }
#'
#' @examples ## See README.md on GitHub for a comprehensive example.
#'
#' @docType package
#' @name kcopula
NULL
